perception/sandbox/eyehandcal/tests/test_cal.py

```diff
@@ -167,3 +163,23 @@ def makepose(rotvec, trans):
     plt.savefig('camera_pose.pdf')
 
 
+from eyehandcal.scripts import collect_data_and_cal
+testcases = [
+    [f'--datafile={os.path.join(localpath,"caldata_world_cam.pkl")}', '--marker-id=0'],
+    [f'--datafile={os.path.join(localpath,"caldata_wrist_cam.pkl")}', '--marker-id=1', '--proj-func=world_marker_proj_hand_camera'],
+]
+@pytest.mark.parametrize("argv", testcases, ids = [x[0] for x in testcases])
+def test_collect_data_and_cal(argv):
+    collect_data_and_cal.main(argv)
+    with open('calibration.json', 'rb') as f:
+        results = json.load(f)
+        for result in results:
+            assert result['pixel_error'] < 2.0
```

Comment:
I noticed I can (surprisingly) consistently get <1px error, even with as few as 7 images with visible markers, as long as I have enough retries (might take 10-15min). Is 2.0 too lenient? Having a good calibration allows us to get very accurate pointcloud merging with no effort.

Reply:
My experience is once that stage 1 gives < 50px error, stage almost always converge to the same spot (ie,. same error). And continuous running doesn't yield any better result. I am not sure if your experience is with the original randomized solver or with the SolvePnP stage1 solver.

I increase from 1px to 2 mostly because I can't get <1px solution with some RoboPen camera. Maybe the frame is wobbly or so that needs more time to confirm.


To avoid overfitting, the correct way of evaluate pixel_error will be looking at test_error instead of training error reported here. Having said so, have small training error is still good. Let's leave these tweak as future work.

Reply:
After using the 2stage in main, my hard cases converge rapidly. Thanks.